lib/stages/DotProduct.cpp

```diff
@@ -0,0 +1,107 @@
+#include "DotProduct.hpp"
+
+#include "StageFactory.hpp"
+#include "errors.h"
+#include "visUtil.hpp" // for frameID, modulo
+
+/* Load the classes we will be using
+ * TODO: Annotate the includes and usings */
+
+using kotekan::bufferContainer;
+using kotekan::Config;
+using kotekan::Stage;
+
+// Register the stage with abstract factory.
+// Needed for Kotekan to be able to find the stage.
+REGISTER_KOTEKAN_STAGE(DotProduct);
+
+/* Constructor for the stage
+    TODO: Annotate the arguments everywhere
+    Note, that you can use the macro STAGE_CONSTRUCTOR(DotProduct)
+    if your constructor does not need additional customisation
+    and you wish to hide the complexity. */
+DotProduct::DotProduct(Config& config, const std::string& unique_name,
+                       bufferContainer& buffer_container) :
+    Stage(config, unique_name, buffer_container, std::bind(&DotProduct::main_thread, this)) {
+
+    // Register as consumer of in_a_buf and in_b_buf
+    in_a_buf = get_buffer("in_a_buf");
+    register_consumer(in_a_buf, unique_name.c_str());
+
+    in_b_buf = get_buffer("in_b_buf");
+    register_consumer(in_b_buf, unique_name.c_str());
+
+    // Register as a producer of out_buf
+    out_buf = get_buffer("out_buf");
+    register_producer(out_buf, unique_name.c_str());
+
+    // Get the no. of elements in each frame
+    _num_elements = in_a_buf->frame_size / sizeof(float);
+
+    // Ensure the input and output buffers are the same
+    // length for the dot product
+    assert(in_a_buf->frame_size == in_b_buf->frame_size
```

Comment:
Use an `if` statement here, and then throw an `std::runtime_error` with a message if the size rules fail.